From 46c7405df7de8deb97229eacebcee96d61415f3f Mon Sep 17 00:00:00 2001
Message-Id: <46c7405df7de8deb97229eacebcee96d61415f3f.1652325304.git.geliang.tang@suse.com>
From: Vasily Gorbik <gor@linux.ibm.com>
Date: Wed, 12 May 2021 19:42:10 +0200
Subject: [PATCH] objtool: Fix elf_create_undef_symbol() endianness
Git-commit: 46c7405df7de8deb97229eacebcee96d61415f3f
Patch-mainline: v5.13-rc2
References: bsn#258
Modified-by-SEL: Yes, refreshed due to context changes


Currently x86 cross-compilation fails on big endian system with:

  x86_64-cross-ld: init/main.o: invalid string offset 488112128 >= 6229 for section `.strtab'

Mark new ELF data in elf_create_undef_symbol() as symbol, so that libelf
does endianness handling correctly.

Fixes: 2f2f7e47f052 ("objtool: Add elf_create_undef_symbol()")
Signed-off-by: Vasily Gorbik <gor@linux.ibm.com>
Signed-off-by: Ingo Molnar <mingo@kernel.org>
Acked-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
Link: https://lore.kernel.org/r/patch-1.thread-6c9df9.git-d39264656387.your-ad-here.call-01620841104-ext-2554@work.hours
Signed-off-by: Geliang Tang <geliang.tang@suse.com>
---
 tools/objtool/elf.c |    1 +
 1 file changed, 1 insertion(+)

--- a/tools/objtool/elf.c
+++ b/tools/objtool/elf.c
@@ -670,6 +670,7 @@ static int elf_add_string(struct elf *el
 	data->d_buf = str;
 	data->d_size = strlen(str) + 1;
 	data->d_align = 1;
+	data->d_type = ELF_T_SYM;
 
 	len = strtab->len;
 	strtab->len += data->d_size;
